README.md
# Husky

Companion app for X32.

Run the following: `python3 main.py`

For Macs, run `./pyinstller.sh` to create `Husky.app` within the `dist` directory.

## config.py

This is the file that contains the config for your particular setup. These are the particular settings in the file:

### osc

These are the default IP Addresses for `foh` and `iem` mixers. Note that the port is fixed to connect to 10023, which is the X32 default OSC port.

### atemPort

This is the incoming port to [AtemOSC](http://www.atemosc.com). Currently, this only supports a local instance of AtemOSC.

### midi

These are the default MIDI ports to connect to for each particular category of midi commands. Note that `Husky` is a virtual MIDI port that is opened by this application that can be used by all programs on the computer running this app.

#### default

This is the name of the default port to connect to.

#### type

This is the type of midi call to make. Valid options currently are: `cc` and `note`.

#### defaultChannel

This is the default channel to send midi commands on. Should be a number between 1-16.

### serverMidi

These are the MIDI ports to listen to. MIDI commands sent from these sources are able to trigger cue and fader commands. Under `serverMidi` is a dictionary, where the keys are the default souces to listen to and the value of each key is a array of default commands to listen to. The options for each command is:

#### midi

Incoming midi to listen to. Specify `type` (either `Control Change` or `Note`),  the `channel` (number between 1 and 16) and `control` (number between 0 and 127) to listen to.

#### command

Command to trigger if the conditions for a incoming midi is met. Specify `type` (either `Cue` or `Fader`), `page` (page letter or `CURRENT`) and `index` (number between 1-10 for `Cue` and 1-4 for `Fader`). Additionally if `CURRENT` is selected for `page`, there are additional triggers of `Next Page` and `Prev Page` that can be specified in `index`.

### personal

This is a list of the typical personal that you want to save settings for. For each personal, you can specify `channels` and `iem_bus`. These are the target channel and in ear bus for the particular person. If you leave one blank, then it will not save down the particular settings for that person.

### cues

These are the fields related to the cue tab.

Specify `cuePages` to indicate how many pages of cues to have.

Specify `faderPages` to indicate how many pages of faders to have.

#### cueOptions

This contains a dictionary, where the keys are the default columns for the cue page. Under each key, is another dictionary, where the keys are the options for the column and the key is a array of `osc` or `midi` commands in the format specified below in the `osc` file format. A `RESET` option can be specified, which will be triggered when the reset button is clicked.

### faders

This is a list of default settings to load into the faders. The commands are in the following format:

```
[foh|iem] [command] [min float value] [max float value]
```

or for MIDI cues (note that fader only makes sense for control change commands)

```
midi [audio] [channel] [control]
```

On fader change, a corrosponding command will be fired. For OSC commands, this would send a float value between min and max arguement. For MIDI commands, this would send MIDI value equal to the slider position.

Additionally, the field `oscFeedback` can be specified, to specify certain feedback back to the X32 mixer user specified knobs. This is something that you want to explicitly specify if you configure your FOH X32 mixer to send midi commands to this app.

Additionally, specify the `defaultValue` of a fader, to force the fader to start at a certain position. This value should be a number between 0 and 127.

### talkback

Different values related to talkback.

#### destination

Which destination to use for IEM talkback enable/disable. One of either `A` or `B`.

#### channel

This is the channel that FOH talkback is being sent through. Applicable because FOH Talkback Channel is sent to IEM Mixer. Should be in a format similar to `/ch/01`.

#### link

Whether to link buttons on the FOH mixer with the mute of the talkback channel in the IEM mixer. Only applicable for 2 mixer application.

### selectLink

If configured, will send MIDI note on select. Useful for shortcutting selecting channels in linked Ableton section.

#### targetDestination

Name of the target MIDI channel destination.

#### midiChannel

Channel to send to.

### luckyAutoMix

These are default params for the auto-mixer. This is not required, as there have been default parameters that have been hard coded. Parameters include:

#### bus

This is the default bus to use the auto mixer with. The way that this auto mixer works, is that it changes the bus sends for individual enabled channels. Because of this, it is required for the bus sends to be set as **Post Fader**.

#### postFader

This is a boolean value, specifying whether or not we want to base the auto mixer off the meter value on input, or on output. In practice, the output meter value is calculated using the input meter value, adding the fader position. *Note that the X32 auto mixer bases uses input meters for the auto mixer.*

#### threshold

This is the gate threshold. If mics meter volume are below this threshold, then all mics will be gated down to at least -10db. How this works is that `c = 0` if max meter value is below the threshold, so that all values will then translate to be between range -10db to -30db.

Be careful about adjusting threshold to be not equal to the minimum of 100, because latency issues will mean that if channel volume is ducked, it will take more time to respond.

#### m *and* c

The volume that a channel is auto mixed to depends on the following equation:

```
0.5 + arctan((Channel Volume - Max Channel Volume - c) / m) / 6
```

Effectively, `c in (-24, 0)` is the point at which the curve will hit -10db, and then start to concave back up, and `m in (1, 9)` is the slope of the curve. If you want a more aggressive automixer, then set `m` to be lower, and if you want it to be more smoothed out, set `m` to be higher.

Only exception to channels that won't follow the equation is the channel with the max meter volume, which will have unity send gain.

### resetCommands

This is a dictionary containing commands to be fired on reset call, where the key is the osc address and the value is the argument to be fired. The argument should be in the correct data format.

## Data Directory

Data directory contains files that are created and used by this program. The directory has the following types of files:

*Note that for both file formats, the first line is a header line used by the program.*

### osc

These contain osc commands in the following format:

```
[foh|iem] [command] [str|int|float] [value] [(optional, for floats only) fade time]
```

For float values only, you are also able to optionally provide a 5th arguement to specify how long you want to take to fade in the new value.

You are also able to fire MIDI cues, by putting commands in the following format:

```
midi [audio|video|light] [channel:1-16] [control:0-127] [value:0-127]
```

You are also able to open files using the following command:

```
open [filepath]
```

Additionally, you are also able to make changes to the "I'm Feeling Lucky" local auto mixer in the following format:

```
lucky [/ch/01, /ch/02, ...] [assignment: OFF, A, B, ...] [weight: -12.0 - 12.0] 
```

### cue

This is a file to save down the cues to be files. It contains lines in the following format:

```
[key:A-G] [vocalLead:1-4] [snippet filename]
```

## Shortcuts

### Cues

There are 3 different options to fire cues:

1. Press the "Fire" button next to the cue
2. Use cmd+\<cue number\> to fire that numbered cue (Note that Cue 10 is fired by cmd+0. Also cmd+\<page name\> will change pages)
3. Send a MIDI cue, as configured in the `MIDI Input` page in the connection tab.

## Dependencies:

- [python](https://www.python.org/downloads/)
- [pip](https://pip.pypa.io/en/stable/installation/)
- [pyinstaller](https://pypi.org/project/pyinstaller/)
- [python-osc](https://pypi.org/project/python-osc/)
- [mido](https://mido.readthedocs.io/en/latest/installing.html)
- [PyQt6](https://pypi.org/project/PyQt6/)

### For Mac run the following commands:

Install Homebrew: https://brew.sh

Then after completing the post completion steps of homebrew:

```
brew install git
brew install python
python3 -m ensurepip --upgrade
sudo -H pip3 install --prefix=/usr/local pyinstaller
pip3 install -r requirements.txt
```

## X32 OSC Commands Reference File

[pdf](https://wiki.munichmakerlab.de/images/1/17/UNOFFICIAL_X32_OSC_REMOTE_PROTOCOL_%281%29.pdf)

CHANGELOG.md
# Change Logs

### Format of versioning:

```
v[Major Version].[Minor Version].[Patch Version]
```

A **Major Version** iteration means some sort of backwards *incompatible* change has been made.

A **Minor Version** iteration means some sort of backwards *compatible* major feature has been added.

A **Patch Version** iteration means some sort of backwards *compatible* minor feature or bug fix has been made.

(See https://semver.org)

## v1.0.0 (2022-12-27)

First version of Husky!

## v1.1.0 (2023-03-05)

Add Preferences Menu, that allows you to modify the config object. This is a catch all preferences menu,
to be able to modify preferences for anything.

Also fundamentally changes the way that cache works, to save all defaults in the config.cache file.

## v1.2.0 (2023-03-21)

Add Channel Swap feature in Menu.

## v1.3.0 (2023-03-22)

Add "I'm feeling lucky" auto mixer.

## v2.0.0 (2023-03-30)

Add Talkback link, where FOH talkback buttons are linked with the mute/unmute of the talkback channel on the IEM mixer.

This is backwards incompatible change, as it changes the format of the config. Talkback config is now in the following format:

```
    "talkback": {
        "destination": "B",
        "channel": "/ch/30",
        "link": true
    },
```

Requires manual change in existing instances of Husky. Only really runs on 2 computers right now, so this change is managable.

## v2.1.0 (2023-04-08)

Add ability to use "st" (Main Out) or "mono" (Mono Out) as a iem_bus.

Also changing config to match next evolution of 45th loft.

## v2.2.0 (2023-04-13)

Add ability to open files from osc snippet files.

## v2.3.0 (2023-04-14)

Add ability to fire MIDI command off of mixer select changes.

## v2.3.1 (2023-04-27)

Change MIDI command fired from note_on to control_change

## v2.3.2 (2023-04-27)

Change MIDI command fired from control_change to note_on

## v2.3.3 (2023-04-28)

Fix talkback link

## v2.3.4 (2023-05-03)

Don't transfer link to IEM mixer if talkback channel in the channel pair

## v2.4.0 (2023-05-14)

Add ability to recall MIDI history from all control change commands targeted at Husky.

Useful for recalling History of all outputs from Ableton.

## v2.5.0 (2023-05-23)

Add ability to update MIDI state based on MIDI history.

util/constants.py
APP_NAME = "Husky"
VERSION = "v2.5.0" # Version of the app. TODO for each update, to iterate the version number, as well as iterating the CHANGELOG.

MIXER_TYPE = "X32" # Options: X32, XR18
PORT = 10023 if MIXER_TYPE == "X32" else 10024
START_PORT = 10000 # Ports: 10000 - AtemServer; 10001 - 10010 - bulk_send_message; 10011 - fohServer; 10012 - fohServer.subscription; 10013 - iemServer; 10014 - iemServer.subscription
NUM_THREADS = 10

GAIN_METERS_CMD = "/meters/1"
GAIN_METERS_EXPECTED_FLOATS = 96

AUTOMIX_METERS_CMD = "/meters/4"
AUTOMIX_METERS_EXPECTED_FLOATS = 82

SELECT_STAT = "/-stat/selidx"
TALKBACK_STAT_PREFIX = "/-stat/talk/"

ALL_CHANNELS = [
    "/ch/01", "/ch/02", "/ch/03", "/ch/04", "/ch/05", "/ch/06", "/ch/07", "/ch/08",
    "/ch/09", "/ch/10", "/ch/11", "/ch/12", "/ch/13", "/ch/14", "/ch/15", "/ch/16",
    "/ch/17", "/ch/18", "/ch/19", "/ch/20", "/ch/21", "/ch/22", "/ch/23", "/ch/24",
    "/ch/25", "/ch/26", "/ch/27", "/ch/28", "/ch/29", "/ch/30", "/ch/31", "/ch/32",
    "/auxin/01", "/auxin/02", "/auxin/03", "/auxin/04", "/auxin/05", "/auxin/06", "/auxin/07", "/auxin/08",
    "/fxrtn/01", "/fxrtn/02", "/fxrtn/03", "/fxrtn/04", "/fxrtn/05", "/fxrtn/06", "/fxrtn/07", "/fxrtn/08"
]
AUX_CHANNELS = [
    "/auxin/01", "/auxin/02", "/auxin/03", "/auxin/04", "/auxin/05", "/auxin/06", "/auxin/07", "/auxin/08",
    "/fxrtn/01", "/fxrtn/02", "/fxrtn/03", "/fxrtn/04", "/fxrtn/05", "/fxrtn/06", "/fxrtn/07", "/fxrtn/08"
]
HEADAMP_CHANNELS = {
    "Local": ["{:03d}".format(idx) for idx in range(0, 32)],
    "AES-A": ["{:03d}".format(idx) for idx in range(32, 80)],
    "AES-B": ["{:03d}".format(idx) for idx in range(80, 128)]
}
LINK_CHANNELS = [
    "1-2", "3-4", "5-6", "7-8", "9-10", "11-12", "13-14", "15-16",
    "17-18", "19-20", "21-22", "23-24", "25-26", "27-28", "29-30", "31-32"
]

ALL_BUSES = ["01", "02", "03", "04", "05", "06", "07", "08", "09", "10", "11", "12", "13", "14", "15", "16"]
ODD_BUSES = ["01", "03", "05", "07", "09", "11", "13", "15"]

ALL_MATRICIES = ["01", "02", "03", "04", "05", "06"]
ODD_MATRICIES = ["01", "03", "05"]

BANKS_32 = ["1-8", "9-16", "17-24", "25-32"]
BANKS_48 = ["1-8", "9-16", "17-24", "25-32", "33-40", "41-48"]
BANKS_16 = ["1-4", "5-8", "9-12", "13-16"]
ROUTING_IN = [
    "Local XLR In 1-8", "Local XLR In 9-16", "Local XLR In 17-24", "Local XLR In 25-32",
    "AES-A In 1-8", "AES-A In 9-16", "AES-A In 17-24", "AES-A In 25-32", "AES-A In 33-40", "AES-A In 41-48",
    "AES-B In 1-8", "AES-B In 9-16", "AES-B In 17-24", "AES-B In 25-32", "AES-B In 33-40", "AES-B In 41-48",
    "Card In 1-8", "Card In 9-16", "Card In 17-24", "Card In 25-32",
    "User In 1-8", "User In 9-16", "User In 17-24", "User In 25-32",
]
ROUTING_IN_AUX = [
    "Local AUX In 1-6",
    "Local XLR In 1-2", "Local XLR In 1-4", "Local XLR In 1-6",
    "AES-A In 1-2", "AES-A In 1-4", "AES-A In 1-6",
    "AES-B In 1-2", "AES-B In 1-4", "AES-B In 1-6",
    "Card In 1-2", "Card In 1-4", "Card In 1-6",
    "User In 1-2", "User In 1-4", "User In 1-6",
]
ROUTING_IN_USER = [
    "OFF",
    "Local XLR In 01", "Local XLR In 02", "Local XLR In 03", "Local XLR In 04", "Local XLR In 05", "Local XLR In 06", "Local XLR In 07", "Local XLR In 08",
    "Local XLR In 09", "Local XLR In 10", "Local XLR In 11", "Local XLR In 12", "Local XLR In 13", "Local XLR In 14", "Local XLR In 15", "Local XLR In 16",
    "Local XLR In 17", "Local XLR In 18", "Local XLR In 19", "Local XLR In 20", "Local XLR In 21", "Local XLR In 22", "Local XLR In 23", "Local XLR In 24",
    "Local XLR In 25", "Local XLR In 26", "Local XLR In 27", "Local XLR In 28", "Local XLR In 29", "Local XLR In 30", "Local XLR In 31", "Local XLR In 32",
    "AES-A In 01", "AES-A In 02", "AES-A In 03", "AES-A In 04", "AES-A In 05", "AES-A In 06", "AES-A In 07", "AES-A In 08",
    "AES-A In 09", "AES-A In 10", "AES-A In 11", "AES-A In 12", "AES-A In 13", "AES-A In 14", "AES-A In 15", "AES-A In 16",
    "AES-A In 17", "AES-A In 18", "AES-A In 19", "AES-A In 20", "AES-A In 21", "AES-A In 22", "AES-A In 23", "AES-A In 24",
    "AES-A In 25", "AES-A In 26", "AES-A In 27", "AES-A In 28", "AES-A In 29", "AES-A In 30", "AES-A In 31", "AES-A In 32",
    "AES-A In 33", "AES-A In 34", "AES-A In 35", "AES-A In 36", "AES-A In 37", "AES-A In 38", "AES-A In 39", "AES-A In 40",
    "AES-A In 41", "AES-A In 42", "AES-A In 43", "AES-A In 44", "AES-A In 45", "AES-A In 46", "AES-A In 47", "AES-A In 48",
    "AES-B In 01", "AES-B In 02", "AES-B In 03", "AES-B In 04", "AES-B In 05", "AES-B In 06", "AES-B In 07", "AES-B In 08",
    "AES-B In 09", "AES-B In 10", "AES-B In 11", "AES-B In 12", "AES-B In 13", "AES-B In 14", "AES-B In 15", "AES-B In 16",
    "AES-B In 17", "AES-B In 18", "AES-B In 19", "AES-B In 20", "AES-B In 21", "AES-B In 22", "AES-B In 23", "AES-B In 24",
    "AES-B In 25", "AES-B In 26", "AES-B In 27", "AES-B In 28", "AES-B In 29", "AES-B In 30", "AES-B In 31", "AES-B In 32",
    "AES-B In 33", "AES-B In 34", "AES-B In 35", "AES-B In 36", "AES-B In 37", "AES-B In 38", "AES-B In 39", "AES-B In 40",
    "AES-B In 41", "AES-B In 42", "AES-B In 43", "AES-B In 44", "AES-B In 45", "AES-B In 46", "AES-B In 47", "AES-B In 48",
    "Card In 01", "Card In 02", "Card In 03", "Card In 04", "Card In 05", "Card In 06", "Card In 07", "Card In 08",
    "Card In 09", "Card In 10", "Card In 11", "Card In 12", "Card In 13", "Card In 14", "Card In 15", "Card In 16",
    "Card In 17", "Card In 18", "Card In 19", "Card In 20", "Card In 21", "Card In 22", "Card In 23", "Card In 24",
    "Card In 25", "Card In 26", "Card In 27", "Card In 28", "Card In 29", "Card In 30", "Card In 31", "Card In 32",
    "Local AUX In 1", "Local AUX In 2", "Local AUX In 3", "Local AUX In 4", "Local AUX In 5", "Local AUX In 6", "Talkback Internal", "Talkback External",
]
ROUTING_OUT = [
    "Local XLR In 1-8", "Local XLR In 9-16", "Local XLR In 17-24", "Local XLR In 25-32",
    "AES-A In 1-8", "AES-A In 9-16", "AES-A In 17-24", "AES-A In 25-32", "AES-A In 33-40", "AES-A In 41-48",
    "AES-B In 1-8", "AES-B In 9-16", "AES-B In 17-24", "AES-B In 25-32", "AES-B In 33-40", "AES-B In 41-48",
    "Card In 1-8", "Card In 9-16", "Card In 17-24", "Card In 25-32",
    "Out Patch 1-8", "Out Patch 9-16",
    "P16 1-8", "P16 9-16",
    "AUX Channel 1-6/Monitor", "AUX In 1-6/Talkback",
    "User Out 1-8", "User Out 9-16", "User Out 17-24", "User Out 25-32", "User Out 33-40", "User Out 41-48",
    "User In 1-8", "User In 9-16", "User In 17-24", "User In 25-32",
]
ROUTING_OUT_LOCAL_A = [
    "Local XLR In 1-4", "Local XLR In 9-12", "Local XLR In 17-20", "Local XLR In 25-28",
    "AES-A In 1-4", "AES-A In 9-12", "AES-A In 17-20", "AES-A In 25-28", "AES-A In 33-36", "AES-A In 41-44",
    "AES-B In 1-4", "AES-B In 9-12", "AES-B In 17-20", "AES-B In 25-28", "AES-B In 33-36", "AES-B In 41-44",
    "Card In 1-4", "Card In 9-12", "Card In 17-20", "Card In 25-28",
    "Out Patch 1-4", "Out Patch 9-12",
    "P16 1-4", "P16 9-12",
    "AUX Channel 1-4", "AUX In 1-4",
    "User Out 1-4", "User Out 9-12", "User Out 17-20", "User Out 25-28", "User Out 33-36", "User Out 41-44",
    "User In 1-4", "User In 9-12", "User In 17-20", "User In 25-28",
]
ROUTING_OUT_LOCAL_B = [
    "Local XLR In 5-8", "Local XLR In 13-16", "Local XLR In 21-24", "Local XLR In 29-32",
    "AES-A In 5-8", "AES-A In 13-16", "AES-A In 21-24", "AES-A In 29-32", "AES-A In 37-40", "AES-A In 45-48",
    "AES-B In 5-8", "AES-B In 13-16", "AES-B In 21-24", "AES-B In 29-32", "AES-B In 37-40", "AES-B In 45-48",
    "Card In 5-8", "Card In 13-16", "Card In 21-24", "Card In 29-32",
    "Out Patch 5-8", "Out Patch 13-16",
    "P16 5-8", "P16 13-16",
    "AUX Channel 5-6/Monitor", "AUX In 5-6/Talkback",
    "User Out 5-8", "User Out 13-16", "User Out 21-24", "User Out 29-32", "User Out 37-40", "User Out 45-48",
    "User In 5-8", "User In 13-16", "User In 21-24", "User In 29-32",
]
ROUTING_OUT_DIGITAL = [
    "OFF", "Main L", "Main R", "Main M/C",
    "Bus 01", "Bus 02", "Bus 03", "Bus 04", "Bus 05", "Bus 06", "Bus 07", "Bus 08",
    "Bus 09", "Bus 10", "Bus 11", "Bus 12", "Bus 13", "Bus 14", "Bus 15", "Bus 16",
    "Matrix 01", "Matrix 02", "Matrix 03", "Matrix 04", "Matrix 05", "Matrix 06",
    "Channel 01", "Channel 02", "Channel 03", "Channel 04", "Channel 05", "Channel 06", "Channel 07", "Channel 08",
    "Channel 09", "Channel 10", "Channel 11", "Channel 12", "Channel 13", "Channel 14", "Channel 15", "Channel 16",
    "Channel 17", "Channel 18", "Channel 19", "Channel 20", "Channel 21", "Channel 22", "Channel 23", "Channel 24",
    "Channel 25", "Channel 26", "Channel 27", "Channel 28", "Channel 29", "Channel 30", "Channel 31", "Channel 32",
    "AUX Channel 1", "AUX Channel 2", "AUX Channel 3", "AUX Channel 4", "AUX Channel 5", "AUX Channel 6", "AUX Channel 7", "AUX Channel 8",
    "FX Return 1L", "FX Return 1R", "FX Return 2L", "FX Return 2R", "FX Return 3L", "FX Return 3R", "FX Return 4L", "FX Return 4R",
    "Monitor L", "Monitor R", "Talkback Out"
]
ROUTING_OUT_USER = ROUTING_IN_USER + [
    "Out Patch 01", "Out Patch 02", "Out Patch 03", "Out Patch 04", "Out Patch 05", "Out Patch 06", "Out Patch 07", "Out Patch 08",
    "Out Patch 09", "Out Patch 10", "Out Patch 11", "Out Patch 12", "Out Patch 13", "Out Patch 14", "Out Patch 15", "Out Patch 16",
    "P16 01", "P16 02", "P16 03", "P16 04", "P16 05", "P16 06", "P16 07", "P16 08",
    "P16 09", "P16 10", "P16 11", "P16 12", "P16 13", "P16 14", "P16 15", "P16 16",
    "AUX Channel 1", "AUX Channel 2", "AUX Channel 3", "AUX Channel 4", "AUX Channel 5", "AUX Channel 6", "Monitor L", "Monitor R",
]

SETTINGS = {
    "Label": ["/config/name", "/config/icon", "/config/color"],
    "Gate": ["/gate/mode", "/gate/thr", "/gate/range", "/gate/attack", "/gate/hold", "/gate/release"],
    "HPF": ["/preamp/hpon", "/preamp/hpf", "/preamp/hpslope"],
    "EQ": [
        "/eq/1/type", "/eq/1/f", "/eq/1/g", "/eq/1/q",
        "/eq/2/type", "/eq/2/f", "/eq/2/g", "/eq/2/q",
        "/eq/3/type", "/eq/3/f", "/eq/3/g", "/eq/3/q",
        "/eq/4/type", "/eq/4/f", "/eq/4/g", "/eq/4/q"
    ],
    "Dynamics": ["/dyn/thr", "/dyn/ratio", "/dyn/knee", "/dyn/mgain", "/dyn/attack", "/dyn/hold", "/dyn/release", "/dyn/mix"],
    "Pan and Mute": ["/mix/pan", "/mix/on"]
}

ADDITIONAL_SETTINGS = {
    "Delay": ["/delay/on", "/delay/time"],
    "Insert": ["/insert/on", "/insert/pos", "/insert/sel"],
    "Groups": ["/grp/dca", "/grp/mute"],
    "Sends": [
        "/mix/01/on", "/mix/01/level", "/mix/01/type", "/mix/01/pan", "/mix/01/panFollow",
        "/mix/02/on", "/mix/02/level",
        "/mix/03/on", "/mix/03/level", "/mix/03/type", "/mix/03/pan", "/mix/03/panFollow",
        "/mix/04/on", "/mix/04/level",
        "/mix/05/on", "/mix/05/level", "/mix/05/type", "/mix/05/pan", "/mix/05/panFollow",
        "/mix/06/on", "/mix/06/level",
        "/mix/07/on", "/mix/07/level", "/mix/07/type", "/mix/07/pan", "/mix/07/panFollow",
        "/mix/08/on", "/mix/08/level",
        "/mix/09/on", "/mix/09/level", "/mix/09/type", "/mix/09/pan", "/mix/09/panFollow",
        "/mix/10/on", "/mix/10/level",
        "/mix/11/on", "/mix/11/level", "/mix/11/type", "/mix/11/pan", "/mix/11/panFollow",
        "/mix/12/on", "/mix/12/level",
        "/mix/13/on", "/mix/13/level", "/mix/13/type", "/mix/13/pan", "/mix/13/panFollow",
        "/mix/14/on", "/mix/14/level",
        "/mix/15/on", "/mix/15/level", "/mix/15/type", "/mix/15/pan", "/mix/15/panFollow",
        "/mix/16/on", "/mix/16/level",
    ]
}

SETTINGS_BUS_MTX = {
    "Label": ["/config/icon", "/config/color"],
    "EQ": [
        "/eq/1/type", "/eq/1/f", "/eq/1/g", "/eq/1/q",
        "/eq/2/type", "/eq/2/f", "/eq/2/g", "/eq/2/q",
        "/eq/3/type", "/eq/3/f", "/eq/3/g", "/eq/3/q",
        "/eq/4/type", "/eq/4/f", "/eq/4/g", "/eq/4/q",
        "/eq/5/type", "/eq/5/f", "/eq/5/g", "/eq/5/q",
        "/eq/6/type", "/eq/6/f", "/eq/6/g", "/eq/6/q"
    ],
    "Dynamics": ["/dyn/thr", "/dyn/ratio", "/dyn/knee", "/dyn/mgain", "/dyn/attack", "/dyn/hold", "/dyn/release", "/dyn/mix"],
    "Pan and Mute": ["/mix/pan", "/mix/on"]
}

SETTINGS_MAIN = {
    "EQ": [
        "/eq/1/type", "/eq/1/f", "/eq/1/g", "/eq/1/q",
        "/eq/2/type", "/eq/2/f", "/eq/2/g", "/eq/2/q",
        "/eq/3/type", "/eq/3/f", "/eq/3/g", "/eq/3/q",
        "/eq/4/type", "/eq/4/f", "/eq/4/g", "/eq/4/q",
        "/eq/5/type", "/eq/5/f", "/eq/5/g", "/eq/5/q",
        "/eq/6/type", "/eq/6/f", "/eq/6/g", "/eq/6/q"
    ],
    "Dynamics": ["/dyn/thr", "/dyn/ratio", "/dyn/knee", "/dyn/mgain", "/dyn/attack", "/dyn/hold", "/dyn/release", "/dyn/mix"]
}